packages/smart-contracts/test/contracts/EthereumSingleRequestProxy.test.ts

```diff
@@ -0,0 +1,143 @@
+      .withArgs(payeeAddress, paymentAmount, paymentReference, feeAmount, feeRecipientAddress);
+
+    expect(await ethers.provider.getBalance(ethereumSingleRequestProxy.address)).to.equal(0);
+    expect(await ethers.provider.getBalance(ethereumFeeProxy.address)).to.equal(0);
+  });
+
+  it('should handle funds sent back from EthereumFeeProxy', async () => {
+    const MockEthereumFeeProxyFactory = await ethers.getContractFactory('MockEthereumFeeProxy');
+    const mockEthereumFeeProxy = await MockEthereumFeeProxyFactory.deploy();
+    await mockEthereumFeeProxy.deployed();
+
+    const newEthereumSingleRequestProxyFactory = await ethers.getContractFactory(
+      'EthereumSingleRequestProxy',
+    );
+    const newEthereumSingleRequestProxy = await newEthereumSingleRequestProxyFactory.deploy(
+      payeeAddress,
+      paymentReference,
+      mockEthereumFeeProxy.address,
+      feeRecipientAddress,
+      feeAmount,
+    );
+    await newEthereumSingleRequestProxy.deployed();
+
+    const paymentAmount = ethers.utils.parseEther('1');
+    const totalAmount = paymentAmount.add(feeAmount);
+
+    await owner.sendTransaction({
+      to: newEthereumSingleRequestProxy.address,
+      value: totalAmount,
+    });
+
+    expect(await ethers.provider.getBalance(newEthereumSingleRequestProxy.address)).to.equal(0);
+    expect(await ethers.provider.getBalance(mockEthereumFeeProxy.address)).to.equal(totalAmount);
+
+    await expect(() =>
+      mockEthereumFeeProxy.sendFundsBack(newEthereumSingleRequestProxy.address, totalAmount),
+    ).to.changeEtherBalances(
+      [owner, newEthereumSingleRequestProxy, mockEthereumFeeProxy],
+      [totalAmount, 0, totalAmount.mul(-1)],
+    );
+
+    expect(await ethers.provider.getBalance(newEthereumSingleRequestProxy.address)).to.equal(0);
+    expect(await ethers.provider.getBalance(mockEthereumFeeProxy.address)).to.equal(0);
+  });
+
+  it('should rescue funds', async () => {
+    const paymentAmount = ethers.utils.parseEther('1');
+    const totalAmount = paymentAmount.add(feeAmount);
+
+    const ForceSendFactory = await ethers.getContractFactory('ForceSend');
+    const forceSend = await ForceSendFactory.deploy();
+    await forceSend.deployed();
+
+    await forceSend.forceSend(ethereumSingleRequestProxy.address, { value: totalAmount });
+
+    const balanceAfterForceSend = await ethers.provider.getBalance(
+      ethereumSingleRequestProxy.address,
+    );
+    expect(balanceAfterForceSend).to.be.gt(0);
+    expect(balanceAfterForceSend).to.equal(totalAmount);
+
+    const initialPayeeBalance = await ethers.provider.getBalance(payeeAddress);
+
+    await ethereumSingleRequestProxy.rescueFunds();
+
+    expect(await ethers.provider.getBalance(ethereumSingleRequestProxy.address)).to.equal(0);
+
+    const finalPayeeBalance = await ethers.provider.getBalance(payeeAddress);
+    expect(finalPayeeBalance.sub(initialPayeeBalance)).to.equal(balanceAfterForceSend);
+  });
+});
```

Comment:
We're assuming that our reentrancy guard works because it is nearly identical to the OpenZeppelin ReentrancyGuard with the exception that it allows the first reentrancy call from the ERC20FeeProxy.